### ISSUE TITLE
[Feat]: Dnsmasq monitor metric: cache_size should refect real-time cached entries' number

### ISSUE THREAD

**Issue by L3w1s-L1u** [netdata/netdata#14737]
### Problem

I use several DNS resolvers (dnsmasq) in LAN to caching upstream public DNS query results to improve Internet surfing experience for my orgnization. When I try to monitor DNS caching performance of the resolvers using Netdata, I found  the metric "cache_size" didn't work as I expected, nor the way the Netdata offical documents claim to be: 

In this [blog post](https://blog.netdata.cloud/dnsmasq-monitoring/#monitoring-dnsmasq-with-netdata), metric cache_size tends to reflect the number of entries actually cached, which is a very good indicator of whether the cache size setting in dnsmasq config file fits current workload scenario.  However, it seems that this metric is implemented as just reflecting the static cache size setting value in dnsmasq config file. Here is why I think so: 

1. When I change the cache size setting in dnsmasq.conf, the cache_size reading in dashboard chart changed immediately, as shown in below chart snapshot 
![1678954879075](https://user-images.githubusercontent.com/8043079/225556990-d5be5c11-ac38-43f4-a24c-e6be0eff0bdd.png)
you can see I reduced cache size of resolver1(red) and increased cache size of resolver2(blue). The curves changed instantly which it shoudn't if cache size indicates the actual cached entries. If it were actual cached entries the curve should raise slowly as queries going upstreams get cached gradually.

2. When I shut down a reslover and then restarted it (stack in yellow), the cache size should also increase slowly as outgoing queries begin to fill in the cache. But what we can see here is the cache size just popped to a fixed value which I set in dnsmasq config file.
![1678862890046](https://user-images.githubusercontent.com/8043079/225554632-d3c65520-1dd2-4d29-aead-711f5381cedc.png)

### Description

The cache_size metric in dnsmasq collector should really reflect the real-time cached entry number instead of the cache size setting in config file.

### Importance

really want

### Value proposition

A real-time cached entry number combined with other cache metrics has significant meaning in resolver's performance optimization. Monitor the static cache size setting in dnsmasq config gives very limited hints on how the resolver's cache parameters  should be picked.

**Comment by ilyam8** [comment]
Hi, @L3w1s-L1u. Netdata shows whatever Dnsmasq reports. There is no additional processing.

You can query it yourself. From [Dnsmasq docs](https://manpages.debian.org/stretch/dnsmasq-base/dnsmasq.8.en.html#NOTES):

> The cache statistics are also available in the DNS as answers to queries of class CHAOS and type TXT in domain bind. The domain names are cachesize.bind, insertions.bind, evictions.bind, misses.bind, hits.bind, auth.bind and servers.bind. An example command to query this, using the dig utility would be

> dig +short chaos txt cachesize.bind


To query use `dig +short chaos txt cachesize.bind @127.0.0.1`.

**Comment by L3w1s-L1u** [comment]
Hi @ilyam8 , thanks for your reply! I've tried using dig to qurey just as you shown above. So it turns out the Dnsmasq DO report the cachesize setting rather than the actual cached entry number in real-time.

Then I think I just misunderstood the cache_size meaning metioned in [this blogpost](https://blog.netdata.cloud/dnsmasq-monitoring/#monitoring-dnsmasq-with-netdata) or the explaination of cache_size metric in this post is a little bit misleading.

Anyway appreciate your clarification here!

**Comment by ilyam8** [comment]
I checked Dnsmasq source code:
- [Metrics list](https://github.com/imp/dnsmasq/blob/770bce967cfc9967273d0acfb3ea018fb7b17522/src/metrics.c#L19-L24).
- [Statistics available via DNS queries](https://github.com/imp/dnsmasq/blob/770bce967cfc9967273d0acfb3ea018fb7b17522/src/option.c#L5498-L5511).
- [Stats response](https://github.com/imp/dnsmasq/blob/770bce967cfc9967273d0acfb3ea018fb7b17522/src/cache.c#L1625-L1651). Cache size is `daemon->cachesize`.
- `cachesize` is [the same as the `--cache-size` config option](https://github.com/imp/dnsmasq/blob/770bce967cfc9967273d0acfb3ea018fb7b17522/src/option.c#L3105-L3128).
- `--cache-size` is _the size of dnsmasq's cache. The default is 150 names._ ([man dnsmasq](https://linux.die.net/man/8/dnsmasq)).

TL;DR it is not the current number of entries in the cache as you see. I don't see a way to get the current number. I checked and it is not `inserted` - `evicted`.

**Comment by ilyam8** [comment]
You are right, @L3w1s-L1u. The description of the metric in the blogpost is confusing. @sashwathn we need to update it.

---

And it looks like the metric is far from being useful because we don't have the current number of entries in the cache. 

> -c, --cache-size=<cachesize>
> Set the size of dnsmasq's cache. The default is 150 names. **Setting the cache size to zero disables caching**.

Probably we can remove "Cache size" chart and add "Caching status" with dimensions "enabled"/"disabled". I think this is the best value we can get from this metric.

**Comment by ilyam8** [comment]
On the other hand, probably it worth to have cachesize collected, because you can compare Cache performance (hits/misses) before/after changing the value.

**Comment by ilyam8** [comment]
I changed the cache charts order in netdata/go.d.plugin#1125.

- I think "Cache performance" is the most important one so I made it first in the "cache" section.
- Kept "Cache size" chart, but made it last in "cache".

**Comment by ilyam8** [comment]
@L3w1s-L1u the [first line](https://github.com/imp/dnsmasq/blob/770bce967cfc9967273d0acfb3ea018fb7b17522/src/cache.c#L1726-L1727) doesn't contain additional info, it is:

 - cachesize
 - total evicted
 - total inserted

**Comment by L3w1s-L1u** [comment]
@ilyam8 you're right. I was hoping current cache level could be calculated by data given in first few lines, no chance. I'll see if there is another way.
